### PR TITLE
perf: Unify the undo log of all snapshot types

### PR DESCRIPTION
**Pull request by Marwes**
Extracted from #69218 and extended to all the current snapshot types.

Since snapshotting is such a frequent action in the compiler and many of the scopes execute so little work, the act of creating the snapshot and rolling back empty/small snapshots end up showing in perf. By unifying all the logs into one the creation of snapshots becomes significantly cheaper at the cost of some complexity when combining the log with the specific data structures that are being mutated.

Depends on https://github.com/rust-lang-nursery/ena/pull/29

### CODE REVIEW

**Comment by jonas-schievink** [comment]
(feel free to ping me when this builds so I can run perf)

**Comment by Marwes** [comment]
@jonas-schievink  It should work now, only ran `--stage=1` testing which apparently accepted the mistake in importing.

**Comment by Marwes** [comment]
I believe this should be marked as waiting for review? I don't think there is anything else for me to do here at this stage?

**Comment by JohnCSimon** [comment]
@cramertj - moved this to review as per Marwes

**Comment by cramertj** [comment]
r? @nikomatsakis 
cc @eddyb 

(adding because they reviewed https://github.com/rust-lang/rust/pull/69218)

**Comment by Dylan-DPC** [comment]
@Marwes if you can rebase this, we can get it reviewed before the next conflict hits (or at least try to :P )

**Comment by Marwes** [comment]
Rebased, but it still needs the `ena` PR to be merged and released.

**Comment by eddyb** [comment]
Marking as blocked on the basis of https://github.com/rust-lang/rust/pull/69464#issuecomment-605839589:
> it still needs the `ena` PR to be merged and released.

**Comment by nikomatsakis** [comment]
I did a first review of the ena PR yesterday, will come back to this one afterwards.

**Comment by Marwes** [comment]
Rebased and passing test with ena 0.14

File: src/librustc_infer/infer/mod.rs [hunk 1]
```diff
@@ -196,20 +200,220 @@ impl<'tcx> InferCtxtInner<'tcx> {
...
+}
+
+pub(crate) enum UndoLog<'tcx> {
+    TypeVariables(type_variable::UndoLog<'tcx>),
+    ConstUnificationTable(sv::UndoLog<ut::Delegate<ty::ConstVid<'tcx>>>),
+    IntUnificationTable(sv::UndoLog<ut::Delegate<ty::IntVid>>),
+    FloatUnificationTable(sv::UndoLog<ut::Delegate<ty::FloatVid>>),
+}
+
+impl<'tcx> From<sv::UndoLog<ut::Delegate<type_variable::TyVidEqKey<'tcx>>>> for UndoLog<'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Comment on this impl and all the similar ones:

Upcast from a single kind of "undoable action" to the general enum

File: src/librustc_infer/infer/mod.rs [hunk 2]
```diff
@@ -196,20 +200,220 @@ impl<'tcx> InferCtxtInner<'tcx> {
...
+        match undo {
+            UndoLog::TypeVariables(undo) => self.type_variables.reverse(undo),
+            UndoLog::ConstUnificationTable(undo) => self.const_unification_table.reverse(undo),
+            UndoLog::IntUnificationTable(undo) => self.int_unification_table.reverse(undo),
+            UndoLog::FloatUnificationTable(undo) => self.float_unification_table.reverse(undo),
+        }
+    }
+}
+
+pub(crate) struct Logs<'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Comment:

The combined undo log for all the various unification tables. For each change to the storage for any kind of inference variable, we record an `UndoLog` entry in the vector here.

File: src/librustc_infer/infer/mod.rs [hunk 3]
```diff
@@ -196,20 +200,220 @@ impl<'tcx> InferCtxtInner<'tcx> {
...
+    }
+}
+
+impl<'tcx> Snapshots<UndoLog<'tcx>> for Logs<'tcx> {
+    type Snapshot = Snapshot<'tcx>;
+    fn actions_since_snapshot(&self, snapshot: &Self::Snapshot) -> &[UndoLog<'tcx>] {
+        &self.logs[snapshot.undo_len..]
+    }
+
+    fn start_snapshot(&mut self) -> Self::Snapshot {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Comment: 

Why is this unreachable?

File: src/librustc_infer/infer/mod.rs [hunk 4]
```diff
@@ -196,20 +200,220 @@ impl<'tcx> InferCtxtInner<'tcx> {
...
+    {
+        if self.in_snapshot() {
+            self.logs.extend(undos.into_iter().map(UndoLog::from))
+        }
+    }
+}
+
+impl<'tcx> Snapshots<UndoLog<'tcx>> for Logs<'tcx> {
+    type Snapshot = Snapshot<'tcx>;
+    fn actions_since_snapshot(&self, snapshot: &Self::Snapshot) -> &[UndoLog<'tcx>] {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Nit: space before this `fn`, like the rest =)

File: src/librustc_infer/infer/mod.rs [hunk 5]
```diff
@@ -196,20 +200,220 @@ impl<'tcx> InferCtxtInner<'tcx> {
...
+    num_open_snapshots: usize,
+}
+
+impl Default for Logs<'_> {
+    fn default() -> Self {
+        Self { logs: Default::default(), num_open_snapshots: Default::default() }
+    }
+}
+
+impl<'tcx, T> UndoLogs<T> for Logs<'tcx>
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Comment:

The `UndoLogs` trait defines how we undo a particular kind of action (of type `T`). We can undo any action that is convertable into a `UndoLog` (per the `From` impls above).

File: src/librustc_infer/infer/mod.rs [hunk 6]
```diff
@@ -196,20 +200,220 @@ impl<'tcx> InferCtxtInner<'tcx> {
...
+        Self: Sized,
+        J: IntoIterator<Item = T>,
+    {
+        if self.in_snapshot() {
+            self.logs.extend(undos.into_iter().map(UndoLog::from))
+        }
+    }
+}
+
+impl<'tcx> Snapshots<UndoLog<'tcx>> for Logs<'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Comment: 

The `Snapshots` trait defines the interaction with snapshots.

File: src/librustc_infer/infer/mod.rs [hunk 7]
```diff
@@ -196,20 +200,220 @@ impl<'tcx> InferCtxtInner<'tcx> {
...
+    ut::UnificationTable<ut::InPlace<T, &'a mut ut::UnificationStorage<T>, &'a mut Logs<'tcx>>>;
+
+struct RollbackView<'tcx, 'a> {
+    type_variables: type_variable::RollbackView<'tcx, 'a>,
+    const_unification_table: &'a mut ut::UnificationStorage<ty::ConstVid<'tcx>>,
+    int_unification_table: &'a mut ut::UnificationStorage<ty::IntVid>,
+    float_unification_table: &'a mut ut::UnificationStorage<ty::FloatVid>,
+}
+
+impl<'tcx> Rollback<UndoLog<'tcx>> for RollbackView<'tcx, '_> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Comment:

The `Rollback` trait defines how to rollback a particular action.

File: src/librustc_infer/infer/mod.rs [hunk 8]
```diff
@@ -639,10 +843,11 @@ impl<'tcx> InferOk<'tcx, ()> {
 #[must_use = "once you start a snapshot, you should always consume it"]
 pub struct CombinedSnapshot<'a, 'tcx> {
     projection_cache_snapshot: traits::ProjectionCacheSnapshot,
+    undo_snapshot: Snapshot<'tcx>,
     type_snapshot: type_variable::Snapshot<'tcx>,
-    const_snapshot: ut::Snapshot<ut::InPlace<ty::ConstVid<'tcx>>>,
-    int_snapshot: ut::Snapshot<ut::InPlace<ty::IntVid>>,
-    float_snapshot: ut::Snapshot<ut::InPlace<ty::FloatVid>>,
+    const_snapshot: usize,
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Rename to `const_var_len` and add comment:

Stores the number of inference variables that existed at the time the snapshot was created.

File: src/librustc_infer/infer/mod.rs [hunk 9]
```diff
@@ -639,10 +843,11 @@ impl<'tcx> InferOk<'tcx, ()> {
...
+    int_snapshot: usize,
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: as with `const` but `int`

File: src/librustc_infer/infer/mod.rs [hunk 10]
```diff
@@ -639,10 +843,11 @@ impl<'tcx> InferOk<'tcx, ()> {
...
+    float_snapshot: usize,
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: as with `const` but `float`

File: src/librustc_infer/infer/mod.rs [hunk 11]
```diff
@@ -802,11 +1012,24 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+        let inner = &mut *inner;
+        let InferCtxtInner {
+            type_variables,
+            const_unification_table,
+            int_unification_table,
+            float_unification_table,
+            ..
+        } = inner;
+        inner.undo_log.rollback_to(
+            &mut RollbackView {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Why do we implement `Rollback` for `RollbackView` and not `InferCtxtInner`?

File: src/librustc_infer/infer/type_variable.rs [hunk 12]
```diff
@@ -3,19 +3,76 @@ use rustc_middle::ty::{self, Ty, TyVid};
...
 use rustc_data_structures::unify as ut;
 use std::cmp;
 use std::marker::PhantomData;
 use std::ops::Range;
 
-pub struct TypeVariableTable<'tcx> {
-    values: sv::SnapshotVec<Delegate>,
+use rustc_data_structures::undo_log::{Rollback, Snapshots, UndoLogs};
+
+pub(crate) enum UndoLog<'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/type_variable.rs:16]: Comment:

Represents a single undo-able action that affects a type inference variable.

File: src/librustc_infer/infer/type_variable.rs [hunk 13]
```diff
@@ -3,19 +3,76 @@ use rustc_middle::ty::{self, Ty, TyVid};
...
+    EqRelation(sv::UndoLog<ut::Delegate<TyVidEqKey<'tcx>>>),
+    SubRelation(sv::UndoLog<ut::Delegate<ty::TyVid>>),
+    Values(sv::UndoLog<Delegate>),
+}
+
+impl<'tcx> From<sv::UndoLog<ut::Delegate<TyVidEqKey<'tcx>>>> for UndoLog<'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/type_variable.rs:22]: Comment on this and the upcoming impls:

Convert from a specific kind of undo to the more general `UndoLog`

File: src/librustc_infer/infer/type_variable.rs [hunk 14]
```diff
@@ -3,19 +3,76 @@ use rustc_middle::ty::{self, Ty, TyVid};
...
+    }
+}
+
+impl<'tcx> From<Instantiate> for UndoLog<'tcx> {
+    fn from(l: Instantiate) -> Self {
+        UndoLog::Values(sv::UndoLog::Other(l))
+    }
+}
+
+pub(crate) struct RollbackView<'tcx, 'a> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/type_variable.rs]: Comment:

A view onto the underlying type variable tables that permits rollback.

File: src/librustc_infer/infer/type_variable.rs [hunk 15]
```diff
@@ -3,19 +3,76 @@ use rustc_middle::ty::{self, Ty, TyVid};
...
+}
+
+impl<'tcx, 'a> From<&'a mut TypeVariableStorage<'tcx>> for RollbackView<'tcx, 'a> {
+    fn from(l: &'a mut TypeVariableStorage<'tcx>) -> Self {
+        let TypeVariableStorage { eq_relations, sub_relations, values } = l;
+        Self { eq_relations, sub_relations, values }
+    }
+}
+
+impl<'tcx> Rollback<UndoLog<'tcx>> for RollbackView<'tcx, '_> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/type_variable.rs]: Some kind of comment.

File: src/librustc_infer/infer/mod.rs [hunk 16]
```diff
@@ -315,6 +334,7 @@ struct RollbackView<'tcx, 'a> {
     const_unification_table: &'a mut ut::UnificationStorage<ty::ConstVid<'tcx>>,
     int_unification_table: &'a mut ut::UnificationStorage<ty::IntVid>,
     float_unification_table: &'a mut ut::UnificationStorage<ty::FloatVid>,
+    region_constraints: &'a mut RegionConstraintStorage<'tcx>,
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: rename

File: src/librustc_infer/infer/mod.rs [hunk 17]
```diff
@@ -1151,6 +1154,17 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         r
     }
 
+    pub fn probe_full<R, F>(&self, f: F) -> R
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs:946]: This definitely needs a comment -- what is the role of `probe` vs `probe_full`? It seems that `probe_full` saves more information, but when would you want to do that?

File: src/librustc_infer/infer/mod.rs [hunk 18]
```diff
@@ -892,13 +892,18 @@ impl<'tcx> InferOk<'tcx, ()> {
 }
 
 #[must_use = "once you start a snapshot, you should always consume it"]
-pub struct CombinedSnapshot<'a, 'tcx> {
-    undo_snapshot: Snapshot<'tcx>,
+pub struct FullSnapshot<'a, 'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs:704]: Comment: What is the difference between a "full snapshot" and a "combined snapshot"? When would I use one vs the other?

File: src/librustc_infer/infer/type_variable.rs [hunk 19]
```diff
@@ -43,20 +43,7 @@ impl<'tcx> From<Instantiate> for UndoLog<'tcx> {
...
-
-impl<'tcx, 'a> From<&'a mut TypeVariableStorage<'tcx>> for RollbackView<'tcx, 'a> {
-    fn from(l: &'a mut TypeVariableStorage<'tcx>) -> Self {
-        let TypeVariableStorage { eq_relations, sub_relations, values } = l;
-        Self { eq_relations, sub_relations, values }
-    }
-}
-
-impl<'tcx> Rollback<UndoLog<'tcx>> for RollbackView<'tcx, '_> {
+impl<'tcx> Rollback<UndoLog<'tcx>> for TypeVariableStorage<'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/type_variable.rs:46]: I did want to ask why this couldn't be implemented directly on the storage...

File: src/librustc_infer/infer/mod.rs [hunk 20]
```diff
@@ -1681,13 +1683,14 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     /// having to resort to storing full `GenericArg`s in `stalled_on`.
     #[inline(always)]
     pub fn ty_or_const_infer_var_changed(&self, infer_var: TyOrConstInferVar<'tcx>) -> bool {
+        let mut inner = self.inner.borrow_mut();
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs:1686]: Why these changes? In general, increasing the scope of `borrow_mut` seems risky

File: src/librustc_middle/hir/mod.rs [hunk 21]
```diff
@@ -78,8 +78,8 @@ pub fn provide(providers: &mut Providers<'_>) {
         let module = hir.as_local_hir_id(id.to_def_id()).unwrap();
         &tcx.untracked_crate.modules[&module]
     };
-    providers.hir_owner = |tcx, id| tcx.index_hir(id.krate).map[id.index].signature;
-    providers.hir_owner_items =
-        |tcx, id| tcx.index_hir(id.krate).map[id.index].with_bodies.as_ref().map(|items| &**items);
+    providers.hir_owner = |tcx, id| tcx.index_hir(LOCAL_CRATE).map[id].signature;
```
Comment (nikomatsakis) [inline src/librustc_middle/hir/mod.rs]: did all this random stuff come from the rebase?

**Review by nikomatsakis — COMMENTED** [review]
OK, I left a bunch of documentation nits, but reading this is definitely making me wonder if there isn't some simpler way to achieve these goals. I still feel like I'm struggling a bit to have a 'coherent view' of the new ena API, even if I understand it when I look close.

In general, this PR seems like it would be a good candidate for a Major Change Proposal.

One thought I had while I was reading: maybe we should move the undo-logs to an `Arc<UndoLog>` or something, and then you could just have each of the unification tables take a handle to that `Arc` -- they'd probably be parameterized by something like `Arc<impl UndoLog<..>>`  which gives them "their specific view" onto the undo log (or even `Arc<dyn ..>` to avoid monomorphization costs, though that could have a perf impact).

Doing that would avoid at least *some* of the wrapping and re-wrapping I see in this commit.

**Comment by Marwes** [comment]
> One thought I had while I was reading: maybe we should move the undo-logs to an Arc<UndoLog> or something, and then you could just have each of the unification tables take a handle to that Arc -- they'd probably be parameterized by something like Arc<impl UndoLog<..>> which gives them "their specific view" onto the undo log (or even Arc<dyn ..> to avoid monomorphization costs, though that could have a perf impact).

Using a shared undolog would necessitate a `RefCell` around the log which would imply a small penalty on each mutation (even if it is small, there was another PR recently which explicitly removed `borrow_mut` calls for a small but noticeable improvements https://github.com/rust-lang/rust/pull/68694).  That said, adding "CellVec` (https://github.com/rust-lang/rust/pull/68694#issuecomment-583960961) and using that for the undo log would make it cheaply shareable though. (With ena now parameterized over the storage it would be possible to use `CellVec` there as well and remove the `RefCell` in `InferCtxt`, could be a nice win on its own)

I don't see that monomorphization is a problem here, practically, all methods will only be instantiated once since there is just the one log. (`*Storage` types only calls `default` and the mutations are only called on one type using `&mut InferCtxtLogs`). Regardless of whether `dyn` is used for the log, `ena`s types will still be parameterized by the log so I don't think that changes much.

**Comment by nikomatsakis** [comment]
Yeah, I think the goal for something like `Arc<dyn...>` would just be to keep the types simpler and more readable.

**Comment by nikomatsakis** [comment]
Hmm, at some point I created a `CellVec`, not sure if ever got added or even got removed in the meantime. The idea here would that the only operations you need are to push/pop undo operations? We do sometimes create slices we can iterate over though.

In any case, I'd be curious to know the answers to some of the questions I raised, too, while 

* full vs partial snapshot
* if we can get rid of some of the intermediate types that just hold `&mut` references

**Comment by Marwes** [comment]
> Hmm, at some point I created a CellVec, not sure if ever got added or even got removed in the meantime. The idea here would that the only operations you need are to push/pop undo operations? We do sometimes create slices we can iterate over though.

I believe the idea is that the `CellVec` never yields any references from it (only cloned values). So it can then be freely reallocated and have its values updated via `&` references.

File: src/librustc_infer/infer/mod.rs (hunk 17 above)
Comment (Marwes) [reply src/librustc_infer/infer/mod.rs:946]: I can try to dig into why, but I just noticed that there was only a few `probe` calls that actually used the extra fields that are now in `FullSnapshot` all other calls to `start_snapshot` ignored them so it seemed a waste to record them.

File: src/librustc_infer/infer/mod.rs (hunk 11 above)
Comment (Marwes) [reply src/librustc_infer/infer/mod.rs]: Since the `undo_log` is part of `InferCtxtInner` I had to separate the fields to implement `Snapshots`. There doesn't seem to be a point to implement `Snapshots` explicitly though so I could implement the equivalent methods without being restricted to the trait and have it work without `RollbackView`

File: src/librustc_middle/hir/mod.rs (hunk 21 above)
Comment (Marwes) [reply src/librustc_middle/hir/mod.rs]: Yeah, reviewing commit by commit will give some issues since I had to fix/revert some things like this in the last commit :(

File: src/librustc_infer/infer/mod.rs (hunk 20 above)
Comment (Marwes) [reply src/librustc_infer/infer/mod.rs:1686]: Seemed like a simplification to me, the function itself doesn't recurse, it only checks the type/const variable so there is no way `borrow_mut` may be called again.

File: src/librustc_infer/infer/mod.rs (hunk 8 above)
Comment (Marwes) [reply src/librustc_infer/infer/mod.rs]: Fixed in a later commit

**Comment by nikomatsakis** [comment]
@Marwes Yes. But one of the things you can't easily do is iterate -- well, you can create an iterator, it just needs to use indexing.

**Comment by Marwes** [comment]
I don't think there is anything that really needs indexing though? Iteration + efficient skipping on the iterator should be enough from what I can tell.

**Comment by nikomatsakis** [comment]
@Marwes I'm not sure what you're saying, but regardless the point is that iteration in "ordinary vectors" relies on having references into the data -- if you want to build an iterator over a `CellVec`, it needs to be based on indexing.

e.g., it looks like

```rust
struct CellVecIterator<'me, T> {
   v: &'me CellVec<T>,
   index: usize,
}

impl<T: Clone> Iterator for CellVecIterator<'_, T> {
    type Item = T;

    fn next(&self) -> Self::Item {
        let i = self.index;
        self.index += 1;
        self.v.get(i)
    }
}
```

**Comment by Marwes** [comment]
I know. That isn't an issue for using a `CellVec` for `undo_log`, it does return a slice of the actions but that could be replaced by an iterator is what I meant.

It will cause some problems if `CellVec` were to be used for `SnapshotVec` which relies on `Index/IndexMut`, could be solved with more abstractions though).

Everything other than sharing a reference counted `UndoLog` to avoid the wrapping (and whether to use `CellVec` in that) is fixed at least.

File: src/librustc_infer/infer/mod.rs [hunk 22]
```diff
@@ -636,18 +700,22 @@ impl<'tcx> InferOk<'tcx, ()> {
...
-    projection_cache_snapshot: traits::ProjectionCacheSnapshot,
-    type_snapshot: type_variable::Snapshot<'tcx>,
-    const_snapshot: ut::Snapshot<ut::InPlace<ty::ConstVid<'tcx>>>,
-    int_snapshot: ut::Snapshot<ut::InPlace<ty::IntVid>>,
-    float_snapshot: ut::Snapshot<ut::InPlace<ty::FloatVid>>,
+pub struct FudgeSnapshot<'a, 'tcx> {
+    snapshot: CombinedSnapshot<'a, 'tcx>,
     region_constraints_snapshot: RegionSnapshot,
-    region_obligations_snapshot: usize,
+    type_snapshot: type_variable::Snapshot<'tcx>,
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs:708]: So, this type sounds like it encodes a lot of information, but really it just encodes a length, and the same is true for `RegionSnapshot` -- so essentially this `FudgeSnapshot` is just adding in the length information? 

Is there any data to suggest that storing 5 or 6 integers meaningfully changes performance? If not, it seems to me like it's not worth adding the `probe` vs `probe_fudge` distinction.

**Comment by nikomatsakis** [comment]
Thanks @Marwes, left a question for you.

File: src/librustc_infer/infer/mod.rs (hunk 22 above)
Comment (Marwes) [reply src/librustc_infer/infer/mod.rs:708]: > So, this type sounds like it encodes a lot of information, but really it just encodes a length, and the same is true for RegionSnapshot -- so essentially this FudgeSnapshot is just adding in the length information?

Yes, should perhaps refactor that a bit to make it clear it is just the length.

> Is there any data to suggest that storing 5 or 6 integers meaningfully changes performance? If not, it seems to me like it's not worth adding the probe vs probe_fudge distinction.

I don't have any specific data, and I doubt it makes a large difference. However each individual snapshot that was captured before is just capturing two integers and incrementing a third so these 5-6 integers is roughly as expensive as two snapshots. Since there is evidently gain in reducing how many snapshots are done, doing this relatively small change to reduce them a bit further seemed worthwhile to me (snapshots are done a lot!).

`probe_fudge` is actually only called in the entire compiler so it is definitely a special case, arguably in simplifies things for all the other `probe` uses which do not care about this.

Comment (nikomatsakis) [reply src/librustc_infer/infer/mod.rs:708]: Well, it doesn't really change anything for the callers, right, since they can't even access those fields in any particular way? Indeed, `probe_fudge` could be private to the `infer` module, I think?

I guess the main change is that it adds this extra concept, a `FudgeSnapshot`, and another method, `probe_fudge`, which then means that users have to think about whether that is something that they might want. Privacy helps here some, though not that much.

It's true that snapshots occur a lot, but I assume that their associated costs don't come so much from the actual number of bytes in the snapshot structure.

Comment (Marwes) [reply src/librustc_infer/infer/mod.rs]: > Well, it doesn't really change anything for the callers, right, since they can't even access those fields in any particular way? Indeed, probe_fudge could be private to the infer module, I think?

True, moved it into the fudge module. Also extracted these lengths fields into a struct separate from snapshotting since they do not not truly have anything to do with it.

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc_infer/infer/mod.rs (hunk 22 above)
Comment (nikomatsakis) [reply src/librustc_infer/infer/mod.rs]: OK, yes, I like this reasonably well. Thanks @Marwes .

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
OK, r=me with rebase.

@bors delegate=Marwes

**Comment by mati865** [comment]
@Marwes you should use `@bors r=nikomatsakis`.

**Comment by Dylan-DPC** [comment]
don't worry, got your back xD

@bors r=nikomatsakis

**Comment by ecstatic-morse** [comment]
@Marwes This slows down clean check builds of `inflate` and `keccak` by [a little under 10%](https://perf.rust-lang.org/compare.html?start=43271a39adc26d29350d4830b594c6435472815e&end=8da5869fb738e51438dd1e0697c1b2c84eb11c59&stat=instructions%3Au). Any idea why?

**Comment by nikomatsakis** [comment]
Hmm! I don't think we saw that on the initial round of perf runs, right?

**Comment by Marwes** [comment]
Must have made a mistake in one of the rebases. I think this is the cause https://github.com/rust-lang/rust/pull/71996 .
